### PR TITLE
feat(plugins): Add `@ExtensionConfiguration` annotation, removing `@SpinnakerExtension` and `ConfigurableExtension` interface

### PR DESCRIPTION
**Pull request by jonsie**
*Breaking change:*

Building off work @robzienert did to add constructor injection, this simplifies the story around configuring extensions.  We now annotate configuration classes with `@ExtensionConfiguration`, provide an annotation value that lines up with our config, and things should Just Work.  

The constructor injection will fail if the constructor contains anything other than an `@ExtensionConfiguration` class or a `PluginsSdkImpl`.  I also added the (currently unused) concept of a plugin configuration, if we at some point want a high level config for a plugin.

### CODE REVIEW

File: kork-plugins/src/main/kotlin/com/netflix/spinnaker/kork/plugins/SpinnakerExtensionFactory.kt [hunk 1]
```diff
@@ -68,10 +63,22 @@ class SpinnakerExtensionFactory(
...
+            configFactory.createExtensionConfig(
+              paramType,
+              paramType.getAnnotation(ExtensionConfiguration::class.java).value,
+              pluginWrapper?.descriptor?.pluginId
+            )
+          }
+          else -> {
+            throw IntegrationException("'${extensionClass.simpleName}' extension has unsupported " +
+              "constructor argument class '${paramType.simpleName}'.  Expected argument classes " +
+              "should be annotated with @ExpectedConfiguration or implement PluginSdks.")
```
Comment (jonsie) [inline kork-plugins/src/main/kotlin/com/netflix/spinnaker/kork/plugins/SpinnakerExtensionFactory.kt]: While writing this message I thought - should we have configuration classes implement a config interface?  With that mechanism we would be able to load config classes without requiring the `@ExtensionConfiguration` annotation.  The benefit being one would not have to think about the YAML config structure when writing a config class if the defaults member values suffice. Ya dig?

File: kork-plugins-api/src/main/java/com/netflix/spinnaker/kork/plugins/api/ExtensionConfiguration.java [hunk 2]
```diff
@@ -22,20 +22,23 @@
...
-@Extension
 @Retention(RetentionPolicy.RUNTIME)
 @Target(ElementType.TYPE)
 @Documented
-public @interface SpinnakerExtension {
-  /** The unique canonical id of the extension. */
-  String id();
+public @interface ExtensionConfiguration {
+  @Nonnull
+  String value();
```
Comment (robzienert) [inline kork-plugins-api/src/main/java/com/netflix/spinnaker/kork/plugins/api/ExtensionConfiguration.java]: Let's add some docs on what the the value actually is supposed to be. There's currently no indicator that this value is supposed to be the configuration key (path?)

File: kork-plugins/src/main/kotlin/com/netflix/spinnaker/kork/plugins/SpinnakerExtensionFactory.kt (hunk 1 above)
Comment (robzienert) [reply kork-plugins/src/main/kotlin/com/netflix/spinnaker/kork/plugins/SpinnakerExtensionFactory.kt]: Personally, I prefer annotations over implementing interfaces / extending classes for use cases like this.

File: kork-plugins/src/main/kotlin/com/netflix/spinnaker/kork/plugins/config/ConfigCoordinates.kt [hunk 3]
```diff
@@ -24,15 +24,31 @@ sealed class ConfigCoordinates {
 /**
```
Comment (robzienert) [inline kork-plugins/src/main/kotlin/com/netflix/spinnaker/kork/plugins/config/ConfigCoordinates.kt]: Side thought: We should probably do more to normalize these config coordinates (`toLowerCase()` and maybe even converting from `camelCase` to `snake-case` automatically for people. This would obviously be a different PR, just putting a thought out there.

Comment (jonsie) [reply kork-plugins/src/main/kotlin/com/netflix/spinnaker/kork/plugins/config/ConfigCoordinates.kt]: Good idea.
